lib/sycamore/sycamore/data/table.py

```diff
@@ -0,0 +1,164 @@
+from collections import OrderedDict
+from dataclasses import dataclass, field
+from typing import Any, Optional
+import xml.etree.ElementTree as ET
+
+from sycamore.data import BoundingBox
+from PIL import Image, ImageDraw
+import numpy as np
+from pandas import DataFrame
+
+
+# This data model is similar to that used by Textract and TableTransformers.
+# It is intended to be a common intermediate representation for a variety of
+# table structure recognition models.
+@dataclass(frozen=True)
+class TableCell:
+    content: str
+    rows: list[int]
+    cols: list[int]
+    is_header: bool = False
+    bbox: Optional[BoundingBox] = None
+    # Model/format specific properties
+    properties: dict[str, Any] = field(default_factory=dict)
+
+
+class Table:
+    def __init__(self, cells: list[TableCell], caption: Optional[str] = None):
+        self.cells: list[TableCell] = cells
+        self.caption = caption
+        self.num_rows = max(max(c.rows) for c in self.cells) + 1
+        self.num_cols = max(max(c.cols) for c in self.cells) + 1
+
+    def __eq__(self, other):
+        if type(other) is not type(self):
+            return False
+
+        if self.num_rows != other.num_rows or self.num_cols != other.num_cols:
+            return False
+
+        # TODO: Yikes! this is expensive. Do the sorting in init or something.
+        if self.sorted_cells() != other.sorted_cells():
+            return False
+
+        return True
+
+    def __hash__(self):
+        return hash((self.num_rows, self.num_cols, self.cells))
+
+    def sorted_cells(self):
+        return sorted(self.cells, key=lambda tc: (min(tc.rows), min(tc.cols)))
+
+    # This algorithm is modified from the TableTransformers code. The conversion to Pandas/CSV is
+    # is necessarily lossy since these formats requires tables to be square and don't support
+    # cells spanning multiple rows/columns.
+    #
+    # One of the main decisions here is how to render cells that span multiple rows and columns,
+    # since they are not nativey supported in CSV. The standard approaches are (1) to duplicate
+    # the cell value in each row/column or (2) to put the value in the first row/column in the cell
+    # and empty values in the other. TableTransformers do the former, The textractor library supports
+    # both, but defaults to the latter. Since our primary use case is preparing context for LLMs,
+    # we speculate that duplication may create confusion, so we default to only displaying a cells
+    # content for the first row/column for which it is applicable. The exception is for header rows,
+    # where we duplicate values to each columnn to ensure that every column has a fully qualified header.
+    def to_pandas(self) -> DataFrame:
+        # Find all row nums containing cells marked as headers.
+        header_rows = sorted(set((row_num for cell in self.cells for row_num in cell.rows if cell.is_header)))
```

Comment:
Does sorting put headers at the top?

Reply:
No, not necessarily. Technically (in html for instance) you can have headers throughout the table -- for example "section" headers for groups of cells. The goal here is to find the set of contiguous header rows at the beginning of the document. We squash these into a header for pandas/csv. Those formats don't support headers elsewhere in the table, so we ignore those for the purposes of conversion.